pkg/plugin/plugin.go

```diff
@@ -17,27 +17,88 @@ limitations under the License.
 package plugin
 
 import (
+	"errors"
+	"fmt"
 	"path"
+	"regexp"
+	"strconv"
 	"strings"
 
-	"github.com/spf13/pflag"
-
-	"sigs.k8s.io/kubebuilder/pkg/model/config"
+	"sigs.k8s.io/kubebuilder/pkg/internal/validation"
 )
 
 const DefaultNameQualifier = ".kubebuilder.io"
 
-type Base interface {
-	// Name returns a DNS1123 label string defining the plugin type.
-	// For example, Kubebuilder's main plugin would return "go".
-	Name() string
-	// Version returns the plugin's semantic version, ex. "v1.2.3".
-	//
-	// Note: this version is different from config version.
-	Version() string
-	// SupportedProjectVersions lists all project configuration versions this
-	// plugin supports, ex. []string{"2", "3"}. The returned slice cannot be empty.
-	SupportedProjectVersions() []string
+// Version is a plugin version containing a non-zero integer and an option stability value
+// that if present identifies a version as not stable to some degree.
+type Version struct {
+	Number    int64
+	Stability string
+}
+
+func (v Version) String() string {
+	s := "v" + strconv.FormatInt(v.Number, 10)
+	if v.Stability != "" {
+		s += "-" + v.Stability
+	}
+	return s
+}
+
+// Validate ensures v contains a number and if it has a stability suffix that it is either "alpha" or "beta".
+func (v Version) Validate() error {
+	if v.Number == 0 {
+		return errors.New("integer value cannot be 0")
+	}
+	if v.Stability != "" && v.Stability != "alpha" && v.Stability != "beta" {
+		return errors.New(`suffix must be "alpha" or "beta"`)
+	}
+	return nil
+}
+
+// verRe defines the string format of a version.
+var verRe = regexp.MustCompile("^(v)?([1-9][0-9]*)(-alpha|-beta)?$")
+
+// ParseVersion parses version into a Version, assuming it adheres to format: (v)?[1-9][0-9]*(-(alpha|beta))?
+func ParseVersion(version string) (v Version, err error) {
+	if version == "" {
+		return v, errors.New("plugin version is empty")
+	}
+
+	// A valid version string will have 4 submatches, each of which may be empty: the full string, "v",
+	// the integer, and the stability suffix. Invalid version strings do not have 4 submatches.
+	submatches := verRe.FindStringSubmatch(version)
+	if len(submatches) != 4 {
+		return v, fmt.Errorf("version format must match %s", verRe.String())
+	}
+
+	// Parse version number.
+	if submatches[2] == "" {
+		return v, errors.New("version must contain an integer")
+	}
+	if v.Number, err = strconv.ParseInt(submatches[2], 10, 64); err != nil {
+		return v, err
+	}
+
+	// Parse stability suffix, if any.
+	v.Stability = strings.TrimPrefix(submatches[3], "-")
+
+	return v, v.Validate()
+}
+
+// Compare returns -1 if v < vp, 0 if v == vp, and 1 if v > vp.
+func (v Version) Compare(vp Version) int {
+	if v.Number == vp.Number {
+		s, sp := v.Stability, vp.Stability
+		if s == sp {
+			return 0
+		}
+		if s == "beta" || (s == "alpha" && sp == "") {
+			return 1
+		}
+	} else if v.Number > vp.Number {
+		return 1
+	}
+	return -1
```

Comment:
I think that it would be the same of :  

 ```suggestion
	return strings.Compare( v.String(), vp.String())
```
PS.: passed in the tests as well.

Reply:
You can't use string comparison on integers unfortunately because their bytes do not equate to their numerical value, ex. `strings.Compare("10", "2") == -1`.

Reply:
good point